### ISSUE TITLE
What does gofumpt want from me?

### ISSUE THREAD

**Issue by Antonboom** [mvdan/gofumpt#79]
Hello!

**example.go**
```golang
package example

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"

	"github.com/jjeffery/stringset"
)

type IHTTPClient interface {
	GetJSON(ctx context.Context, path string, query url.Values, headers map[string]string, unmarshalTemplate interface{}) error
	PostJSON(ctx context.Context, path string, body interface{}, headers map[string]string, unmarshalTemplate interface{}, expectedStatus int) error
	NewRequest(ctx context.Context, method string, relPath string, params url.Values, body io.Reader, headers map[string]string) (*http.Request, error)
	Do(req *http.Request) (*http.Response, error)
}

type IInternalClient interface {
	GetAvailabilityForDate(ctx context.Context, clientID string, dateFrom string, dateTo string, sport string) (map[string]string, error)
	GetAvailabilityForEvents(ctx context.Context, clientID string, events []string) (map[string]string, error)
	GetAvailabilityAndLimits(ctx context.Context, maxWorkers, batchSize int, clientID string, events []string) (map[string]string, error)
	GetDisabledEvents(ctx context.Context, clientID string) (map[string]string, error)
	GetEvents(ctx context.Context, ids []string) (map[string]string, error)
	GetStarredEvents(ctx context.Context, clientID string) (stringset.Set, error)
	GetEventLimits(ctx context.Context, clientID string, events []string) (map[string]string, error)
	GetEventLimitsBatched(ctx context.Context, maxWorkers, batchSize int, clientID string, events []string) (map[string]string, error)
	GetSports(ctx context.Context, clientID string) ([]string, error)
	GetRawSports(ctx context.Context, clientID string) (json.RawMessage, error)
}

type IUsersClient interface {
	GetMarketCategories(ctx context.Context) ([]string, error)
	GetUser(ctx context.Context, clientID string) (string, error)
	GetRawUser(ctx context.Context, clientID string) (json.RawMessage, error)
	GetRawInterfaceState(ctx context.Context) (json.RawMessage, error)
	GetRawCompetitions(ctx context.Context) (json.RawMessage, error)
	GetRawMarketCategories(ctx context.Context) (json.RawMessage, error)
}

```

```
$ gofumpt -d test.go       

$ gofumpt -w test.go

$ golangci-lint --version  
golangci-lint has version v1.28.3 built from (unknown, mod sum: "h1:i6KRBN2WlbtEflyjNe6zDaUjLuPkp/j16x/IOxikVnA=") on (unknown)

$  golangci-lint run test.go
test.go:41: File is not `gofumpt`-ed (gofumpt)
}
```

I see similar errors for other files as well. All errors point to the last line in the file.

**Comment by wimspaargaren** [comment]
Experiencing this issue as well in multiple files. 
**mock.go**
```GOLANG
package mock

import "github.com/stretchr/testify/mock"

// MockClient mock of the X client.
type MockClient struct {
	mock.Mock
}

// Check mock of the Check functionality
func (m *MockClient) Check(input1 string, input2 string) (float32, error) {
	args := m.Called(input1, input2)
	return args.Get(0).(float32), args.Error(1)
}
```

```
$ gofumpt -d mock.go       

$ gofumpt -w mock.go

$ golangci-lint --version  
golangci-lint has version v1.28.1 built from (unknown, mod sum: "h1:yU1ejxsNdqBwSKVFqbBseJfSAhCp7mXwp5CP7+cwCaE=") on (unknown)

$  golangci-lint run test.go
mock.go:20: File is not `gofumpt`-ed (gofumpt)
}
```

Found two additional things:

1. The issue resolves when I adjust the Check function to only `1` input parameter
2. Running golangci-lint v1.27 does not output the issue.

**Comment by mvdan** [comment]
Looks like those are different gofumpt versions. After the commit https://github.com/mvdan/gofumpt/commit/d295aba5a43f4e11105a2481a25159c771381ffb twenty days ago, there's a change that's no longer done by default.

You should file the issue with golangci-lint, because they bundle a specific version of gofumpt and I'm not the one in control of that.

**Comment by mvdan** [comment]
Also, as a general thought, you should make sure you're running the same versions of tools that golangci-lint uses. Otherwise, you'll keep having this issue between different tool versions over and over again. An alternative would be for golangci-lint to just apply the gofumpt fixes directly.

**Comment by Antonboom** [comment]
1) Is it not your responsibility to synchronize the versions of `gofumt` and `gofumt in golangci-lint`? 

2) I have the latest gofumpt installed, and running it with `-w` does not help (it's about `apply the gofumpt fixes directly`). Moreover, gofumpt has no tags, do you suggest that we track your changes by commit?

I also note that if `golangci-lint` and `gofumpt` are vendored together (`tools.go`, for example), then `golangci-lint` cannot be built through `go install` - a compilation error due to the gofumpt linter.

**Comment by mvdan** [comment]
>    Is it not your responsibility to synchronize the versions of `gofumt` and `gofumt in golangci-lint`?

No, I don't participate in the golangci-lint project.

>     I have the latest gofumpt installed, and running it with `-w` does not help (it's about `apply the gofumpt fixes directly`).

Yeah, as said, this is likely because golangci-lint has a version that's a little too old.

>    Moreover, gofumpt has no tags, do you suggest that we track your changes by commit?

You can build or pin any master commit you want, yeah. I don't plan to do tagged releases for now, but I might in the future. They wouldn't really solve this issue here, because you could still have different versions all the same.

> I also note that if `golangci-lint` and `gofumpt` are vendored together (`tools.go`, for example), then `golangci-lint` cannot be built through `go install` - a compilation error due to the gofumpt linter.

I don't get that, sorry. I did change the API here slightly, which is something I try to almost never do, but I assume they integrated gofumpt just before I did that a few weeks ago. They could fix that pretty quickly when they update to the latest version.

> UPD: Or do you suggest using `gofumt -w --extra` and this should fix the problem?

Up to you, really - the extra flag just enables extra rules, as you can see from the README. Those used to be enabled all the time.

**Comment by invidian** [comment]
I also experienced the same and the solution was to downgrade this linter and then apply the suggestion.

**Comment by Antonboom** [comment]
Fixed in golangci-lint [v1.29.0](https://github.com/golangci/golangci-lint/releases/tag/v1.29.0)
